### PR TITLE
Support custom issuer with trailing slash

### PR DESCRIPTION
**Pull request by zinosama**
### Summary
If custom issuer is provided with trailing slash, don't append another trailing slash. 

### Reasoning:
* The iss claim in Auth0 token is in the format of `https://...com/` - users of this library should be able to provide the issuer value as is without having to drop the trailing slash.
* It's inconsistent that protocol prefixing takes into consideration of existing values in provided issuer but trailing slash appending doesn't.

### References
N/A

### Testing

* [X] This change adds unit test coverage
* [X] This change has been tested on the latest version of the platform/language

### Checklist

* [X] I have read the [Auth0 contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
* [X] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
* [X] All existing and new tests complete without errors
* [X] All code quality tools/guidelines in the [CONTRIBUTING documentation](CONTRIBUTING.md) have been run/followed

### CODE REVIEW

File: spec/omniauth/auth0/jwt_validator_spec.rb [hunk 1]
```diff
@@ -133,16 +133,26 @@
     end
 
     context 'different from domain' do
-      let(:jwt_validator) do
-        make_jwt_validator(opt_issuer: 'different.auth0.com')
+      shared_examples_for 'has correct issuer and domain' do
+        let(:jwt_validator) { make_jwt_validator(opt_issuer: opt_issuer) }
+
+        it 'should have the correct issuer' do
```
Comment (joecabezas) [inline spec/omniauth/auth0/jwt_validator_spec.rb]: do not use `should` just name the example as `it 'has the correct issuer' do`

File: spec/omniauth/auth0/jwt_validator_spec.rb [hunk 2]
```diff
@@ -133,16 +133,26 @@
...
+          expect(jwt_validator.issuer).to eq('https://different.auth0.com/')
+        end
+
+        it 'should have the correct domain' do
```
Comment (joecabezas) [inline spec/omniauth/auth0/jwt_validator_spec.rb]: do not use `should` just name the example as `it 'has the correct issuer' do`

File: spec/omniauth/auth0/jwt_validator_spec.rb (hunk 1 above)
Comment (zinosama) [reply spec/omniauth/auth0/jwt_validator_spec.rb]: although i agree with you, the use of `should` follows existing pattern, and i figured it's probably worth staying consistent.

File: spec/omniauth/auth0/jwt_validator_spec.rb (hunk 2 above)
Comment (zinosama) [reply spec/omniauth/auth0/jwt_validator_spec.rb]: ditto

File: spec/omniauth/auth0/jwt_validator_spec.rb (hunk 1 above)
Comment (joecabezas) [reply spec/omniauth/auth0/jwt_validator_spec.rb]: oh my bad not familiar with set pattern in project :)

**Review by lbalmaceda — DISMISSED** [review]
I think the issue that this PR is trying to fix could be fixed just by removing the call to `uri_string` for the issuer when present here https://github.com/auth0/omniauth-auth0/blob/16b584f3c0b3918c6e258d4af27968a06aaf4c35/lib/omniauth/auth0/jwt_validator.rb#L23.

The `uri_string` function should only be used to ensure the domain has that format of "https://and.end.with/". The issuer, if any, should be used as is.

That said, I do second the change to ensure the uri we pass to `uri_string` begins with `https://` and ends with `/`

Can I have @joshcanhelp opinion's on this?

**Comment by zinosama** [comment]
> I think the issue that this PR is trying to fix could be fixed just by removing the call to `uri_string` for the issuer when present here
> 
> https://github.com/auth0/omniauth-auth0/blob/16b584f3c0b3918c6e258d4af27968a06aaf4c35/lib/omniauth/auth0/jwt_validator.rb#L23
> 
> .
> The `uri_string` function should only be used to ensure the domain has that format of "https://and.end.with/". The issuer, if any, should be used as is.
> 
> That said, I do second the change to ensure the uri we pass to `uri_string` begins with `https://` and ends with `/`
> 
> Can I have @joshcanhelp opinion's on this?

Thanks for looking into this @lbalmaceda and i'm happy to make changes on this implementation though I'm not entirely sure what you're proposing. I see that you want to ensure the issuer has a correct format (https and /) but you also want to use the issuer as provided (when provided), so what behavior would you like to see when the provided issuer is malformed?

**Comment by lbalmaceda** [comment]
@zinosama what I suggested is to **keep parsing the domain value** into something that looks like `https://` (just like it was before). I'm not sure about adding the trailing slash though, please do manually check by attempting to log in if anything breaks after adding that trailing slash change you already did for this PR.

But for fixing this bug you are reporting, I'd change this

```
 @issuer = uri_string(options.issuer) if options.respond_to?(:issuer) 
```

into 
```
 @issuer = options.issuer if options.respond_to?(:issuer) 
```

This way, the issuer is respecting what the dev explicitly set, or when null will use the domain with `https://` and hopefully (again, try it!) `/` in the end.

**Comment by zinosama** [comment]
> @zinosama what I suggested is to **keep parsing the domain value** into something that looks like `https://` (just like it was before). I'm not sure about adding the trailing slash though, please do manually check by attempting to log in if anything breaks after adding that trailing slash change you already did for this PR.
> 
> But for fixing this bug you are reporting, I'd change this
> 
> ```
>  @issuer = uri_string(options.issuer) if options.respond_to?(:issuer) 
> ```
> 
> into
> 
> ```
>  @issuer = options.issuer if options.respond_to?(:issuer) 
> ```
> 
> This way, the issuer is respecting what the dev explicitly set, or when null will use the domain with `https://` and hopefully (again, try it!) `/` in the end.

I don't disagree with the notion that this library should *not* format `options.issuer`. However, your proposed change will break existing integrations that rely on this library to prepend `https://` and append `/` to `options.issuer`. Therefore, this is probably something you guys can change in the next major version release. As far as this PR is concerned, I think it introduces a backwards compatible middle ground where we can address this issue with trailing slash in a minor version release.

File: lib/omniauth/auth0/jwt_validator.rb [hunk 3]
```diff
@@ -127,7 +127,8 @@ def json_parse(json)
       def uri_string(uri)
         temp_domain = URI(uri)
         temp_domain = URI("https://#{uri}") unless temp_domain.scheme
-        "#{temp_domain}/"
+        temp_domain = "#{temp_domain}/" unless temp_domain.to_s.end_with?('/')
+        temp_domain.to_s
```
Comment (joshcanhelp) [inline lib/omniauth/auth0/jwt_validator.rb]: Maybe do this first so we don't convert to a string twice?

**Comment by zinosama** [comment]
> @lbalmaceda - Appreciate the initial review here 👍
> 
> @zinosama @joecabezas - Thank you for contribution (and tests!) and talking through the implications. Changes look good for a good balance of backwards-compatibility and usability.
> 
> What I'm curious out is why you're passing in a custom issuer here? Using a custom domain to log in generates an ID token with that custom domain as the issuer. How are you using Omniauth to generate an ID token with a different issuer than the domain you're using to log in?

i'm not sure about the `id token` but the access token i get back has its `iss` set to the auth0 domain (not the custom domain), and it fails the validation in this library. I followed the Auth0 guide on integration for 'regular web app' (rails app in our case), and i think the only thing special about our project is that it has a custom domain.

File: lib/omniauth/auth0/jwt_validator.rb (hunk 3 above)
Comment (zinosama) [reply lib/omniauth/auth0/jwt_validator.rb]: Updated

**Comment by joshcanhelp** [comment]
@zinosama - Now I'm extra confused! Token validation in this library should only happen on the ID token, never the access token. The ID token is validated here:

https://github.com/auth0/omniauth-auth0/blob/master/lib/omniauth/strategies/auth0.rb#L53

Are you validating the access token in your application using this library? If access tokens were getting validated in this library, we should have seen a lot of applications breaking when it was introduced (the access token issued if you're not calling an API isn't even a JWT).

**Comment by joshcanhelp** [comment]
@zinosama - Still curious on the questions above but one more for you ... are you logging in with a social provider without your own development keys? That came up recently as a scenario where the custom domain might not end up in the issuer.

**Comment by zinosama** [comment]
> @zinosama - Now I'm extra confused! Token validation in this library should only happen on the ID token, never the access token. The ID token is validated here:
> 
> https://github.com/auth0/omniauth-auth0/blob/master/lib/omniauth/strategies/auth0.rb#L53
> 
> Are you validating the access token in your application using this library? If access tokens were getting validated in this library, we should have seen a lot of applications breaking when it was introduced (the access token issued if you're not calling an API isn't even a JWT).

Hi @joshcanhelp sorry about the delay in response, and sorry about the confusion - clearly i didn't understand the distinction btw. id token and access token. 

To clarify, you are correct and we are talking id tokens (not access tokens) here. 

The problem, however, is that your statement ` Using a custom domain to log in generates an ID token with that custom domain as the issuer` doesn't seem to be accurate in our case. If it makes any difference, our Auth0 tenant uses the private cloud offering. Please let me know if this is something that I should open a support ticket to investigate.

**Comment by zinosama** [comment]
> @zinosama - Still curious on the questions above but one more for you ... are you logging in with a social provider without your own development keys? That came up recently as a scenario where the custom domain might not end up in the issuer.

I'm not logging in as a social provider. I'm unsure what you mean by our own 'development keys'.

By the way, since our Auth0 tenant configuration is not entirely related to this PR, I don't know if you have a way for us to connect privately so that I can provide more detail specific to our tenant. Let me know if you want me to open a support ticket or send you an email.

**Comment by zinosama** [comment]
@laurawyse @jb08 @pjo336

**Comment by joshcanhelp** [comment]
Apologies on the long delay here. I want to make sure that this use case is validated on our end so we're not supporting something that might, at some point, not be supported. Just a quick note to say that we did not forget about this!

**Review by davidpatrick — APPROVED** [review]

**Pull request merged**